localstack-core/localstack/aws/handlers/validation.py

```diff
@@ -18,16 +19,52 @@
 from localstack.aws.chain import Handler, HandlerChain
 from localstack.constants import INTERNAL_RESOURCE_PATH
 from localstack.http import Response
+from localstack.utils.analytics.usage import UsageSetCounter
 
 LOG = logging.getLogger(__name__)
 
 
+class UsageCollectorFactory:
+    _collector_registry: dict[str, Any] = {}
+    """Registry for the different paths."""
+
+    NAMESPACE_PREFIX = "internal"
+    """Namespace prefix to track usage of public endpoints (_localstack/ and _aws/)."""
+
+    @classmethod
+    def get_collector(cls, path: str):
+        namespace = f"{cls.NAMESPACE_PREFIX}/{path}"
+        if namespace not in cls._collector_registry:
+            cls._collector_registry[namespace] = UsageSetCounter(namespace)
+        return cls._collector_registry[namespace]
+
+
 class OpenAPIValidator(Handler):
     open_apis: list["OpenAPI"]
 
     def __init__(self) -> None:
         self._load_specs()
 
+    def _record_usage(self, context: RequestContext) -> None:
+        if config.DISABLE_EVENTS:
+            return
+
+        request_path = context.request.path
+        user_agent = context.request.headers.get("User-Agent")
+        if not request_path or not user_agent:
+            return
+        # Skip the endpoints for the new API Gateway implementation
+        if "execute-api" in request_path:
+            return
+        # We only record the first segment in the path after the _internal/ or _aws/ prefix, as a path can have
+        # potentially an infinite number of parameters.
+        recorded_path = request_path.split("/")[:2]
+        if len(recorded_path) < 2:
+            return
+        recorded_path = "/".join(recorded_path)
```

Comment:
I think this is a nice compromise for now, while still giving us a picture of which services/features are used in the internal endpoint but not too much granularity/cardinality (especially with dynamic path parameters).

We can still try to think of way to get the "matched" path, maybe with the help of the OpenAPI specs actually? No idea how we could get the "route" like `/_aws/events/rules/{rule_arn}/trigger` for example.

Reply:
Try to get the matched path would be by far the best option, but given the time constraint, I don't see an easy solution we could adopt right now.

Reply:
Definitely agreed! 👍